group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/ModernGroupMember.java

```diff
@@ -0,0 +1,223 @@
+     */
+    protected Set<String> subscribedTopicNames;
+
+    /**
+     * The partitions assigned to this member.
+     */
+    protected Map<Uuid, Set<Integer>> assignedPartitions;
+
+    protected ModernGroupMember(
+        String memberId,
+        int memberEpoch,
+        int previousMemberEpoch,
+        String instanceId,
+        String rackId,
+        int rebalanceTimeoutMs,
+        String clientId,
+        String clientHost,
+        Set<String> subscribedTopicNames,
+        MemberState state,
+        Map<Uuid, Set<Integer>> assignedPartitions
+    ) {
+        this.memberId = memberId;
+        this.memberEpoch = memberEpoch;
+        this.previousMemberEpoch = previousMemberEpoch;
+        this.state = state;
+        this.instanceId = instanceId;
+        this.rackId = rackId;
+        this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+        this.clientId = clientId;
+        this.clientHost = clientHost;
+        this.subscribedTopicNames = subscribedTopicNames;
+        this.assignedPartitions = assignedPartitions;
+    }
+
+    /**
+     * @return The member id.
+     */
+    public String memberId() {
+        return memberId;
+    }
+
+    /**
+     * @return The current member epoch.
+     */
+    public int memberEpoch() {
+        return memberEpoch;
+    }
+
+    /**
+     * @return The previous member epoch.
+     */
+    public int previousMemberEpoch() {
+        return previousMemberEpoch;
+    }
+
+    /**
+     * @return The instance id.
+     */
+    public String instanceId() {
+        return instanceId;
+    }
+
+    /**
+     * @return The rack id.
+     */
+    public String rackId() {
+        return rackId;
+    }
+
+    /**
+     * @return The rebalance timeout in millis.
+     */
+    public int rebalanceTimeoutMs() {
+        return rebalanceTimeoutMs;
+    }
+
+    /**
+     * @return The client id.
+     */
+    public String clientId() {
+        return clientId;
+    }
+
+    /**
+     * @return The client host.
+     */
+    public String clientHost() {
+        return clientHost;
+    }
+
+    /**
+     * @return The list of subscribed topic names.
+     */
+    public Set<String> subscribedTopicNames() {
+        return subscribedTopicNames;
+    }
+
+    /**
+     * @return The current state.
+     */
+    public MemberState state() {
+        return state;
+    }
+
+    /**
+     * @return True if the member is in the Stable state and at the desired epoch.
+     */
+    public boolean isReconciledTo(int targetAssignmentEpoch) {
+        return state == MemberState.STABLE && memberEpoch == targetAssignmentEpoch;
+    }
+
+    /**
+     * @return The set of assigned partitions.
+     */
+    public Map<Uuid, Set<Integer>> assignedPartitions() {
+        return assignedPartitions;
+    }
+
+    /**
+     * @return A boolean indicating whether the member uses the classic protocol, default is false.
+     */
+    public boolean useClassicProtocol() {
+        return false;
+    }
```

Comment:
Does it need to be in the abstract class? The classic protocol can only be used by ConsumerGroupMember, I think.

Reply:
As it's being used in some of the methods in ModernGroup hence needed to define here as well. But these methods are mostly `validateOffsetCommit`, similar offsets method. As in the separate comment discussion, if we move those methods to ConsumerGroup itself with unsupported error in ShareGroup then it ll be much more cleaner. Wdyt?

Reply:
Removed as I moved the code to ConsumerGroup itself.